File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.py
import re

def solution(new_id):
    new_id = new_id.lower() #1

    unusable = re.findall(r'[^-^a-z^0-9^_^.]', new_id) #2
    for u in unusable: new_id = new_id.replace(u, '')
    new_id = new_id.replace('^','')

    while new_id.count('..') > 0: new_id = new_id.replace('..', '.') #3

    if new_id.startswith('.'): #4
        while new_id.startswith('.'): new_id = new_id[1:]
    if new_id.endswith('.'): 
        while new_id.endswith('.'): new_id = new_id[:-1] 

    if new_id == "": new_id = "a" #5

    if len(new_id) >= 16: new_id = new_id[:15] #6
    if new_id.endswith('.'): 
        while new_id.endswith('.'): new_id = new_id[:-1] 

    if len(new_id) <= 2: #7
        c = new_id[-1]
        while len(new_id) <3: new_id += c

    return new_id

File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.js
function solution(new_id) {
    new_id = new_id.toLowerCase();//1단계
    const regex2 = /[a-z0-9-_.]/gm;
    const regex6 = /.{15}/gm;
    let str1 = new_id.match(regex2);//2단계 소문자, 숫자, -, _, .만 남기고 제거
    if(!str1) str1 = "";
    else str1 = str1.join('');
    str1 = str1.replace(/\.{2,}/gm,'.');// 3단계
    str1 = str1.replace(/^\.|\.$/gm,'');// 4단계
    if(str1.length === 0)//5단계 .은 문자열 전체 만약 없으면 if문 안으로
        {
            str1 = 'a';
        }
    if(str1.length >= 16)// 6단계
        {
            // str1 = str1.match(regex6).join('');
            str1 = str1.slice(0, 15);
            str1 = str1.replace(/\.$/gm,'');
        }
    if(str1.length <= 2)//7단계
        {
            let Laststr = str1.match(/.$/gm).join();
            for(let i = str1.length; i < 3; i++)
                {
                    str1 = str1 + Laststr;
                }
        }
    return str1;
}